### DISCUSSION TITLE
Calling Directus Rest API's using the token from Oauth2 server

### DISCUSSION THREAD

**Discussion by phxgbl** [directus/directus#6908] [Feature Requests]
I am using the Keycloak IAM server for Oauth2 authentication. I can login to Directus UI using the same. But i cannot call Rest APIs using the access token generated from the Keycloak server. 
I generated a token using the following command. 

`curl --location --request POST 'http://localhost:9090/auth/realms/Adapt/protocol/openid-connect/token' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'password=1234' \
--data-urlencode 'username=csg-aneer' \
--data-urlencode 'client_id=directus-ui' \
--data-urlencode 'client_secret=ca5a9e89-b35d-4330-8a2b-11cc13a73f63' \
--data-urlencode 'grant_type=password'`

I got the access token. I used the same to call Directus Rest API

`curl --location --request GET 'http://localhost:8055/collections' \
--header 'Authorization: Bearer ACCESS_TOKEN HERE..'`

I got the following error:

`{"errors":[{"message":"Token invalid.","extensions":{"code":"INVALID_CREDENTIALS","stack":"Error: Token invalid.\n    at /home/aneer/workspaces/adapt-wf/directusws/directus-9.0.0-rc.85/api/src/middleware/authenticate.ts:35:11\n    at scoped (/home/aneer/workspaces/adapt-wf/directusws/directus-9.0.0-rc.85/api/src/utils/async-handler.ts:11:21)\n    at Layer.handle [as handle_request] (/home/aneer/workspaces/adapt-wf/directusws/directus-9.0.0-rc.85/node_modules/express/lib/router/layer.js:95:5)\n    at trim_prefix (/home/aneer/workspaces/adapt-wf/directusws/directus-9.0.0-rc.85/node_modules/express/lib/router/index.js:317:13)\n    at /home/aneer/workspaces/adapt-wf/directusws/directus-9.0.0-rc.85/node_modules/express/lib/router/index.js:284:7\n    at Function.process_params (/home/aneer/workspaces/adapt-wf/directusws/directus-9.0.0-rc.85/node_modules/express/lib/router/index.js:335:12)\n    at next (/home/aneer/workspaces/adapt-wf/directusws/directus-9.0.0-rc.85/node_modules/express/lib/router/index.js:275:10)\n    at session (/home/aneer/workspaces/adapt-wf/directusws/directus-9.0.0-rc.85/node_modules/express-session/index.js:479:7)\n    at Layer.handle [as handle_request] (/home/aneer/workspaces/adapt-wf/directusws/directus-9.0.0-rc.85/node_modules/express/lib/router/layer.js:95:5)\n    at trim_prefix (/home/aneer/workspaces/adapt-wf/directusws/directus-9.0.0-rc.85/node_modules/express/lib/router/index.js:317:13)"}}]}`

I copied a token  generated in Directus UI and tried and it worked fine and returned Collections
So my question is: does Directus support calling Rest APIs using Oauth2 provider token directly?

**Comment by rijkvanzanten** [comment]
> So my question is: does Directus support calling Rest APIs using Oauth2 provider token directly?

No, not at this time. Directus only accepts it's own tokens

**Reply by phxgbl** [reply]
@rijkvanzanten  Thank you

**Comment by MarkusElmer** [comment]
When and how got this implemented.
I am still facing the same issue.

**Reply by rijkvanzanten** [reply]
It's not, not sure how that got mis-categorized.

**Reply by MarkusElmer** [reply]
thanks for your quick reply. so i can stop trying.
any plans to integrate that?

**Reply by rijkvanzanten** [reply]
@MarkusElmer Nobody has upvoted this, and you're the second person to bring it up in over a year and a half, so I don't think we're looking to add support for that at this time

**Reply by MarkusElmer** [reply]
ok thx

**Comment by MarkusElmer** [comment]
If somebody else stumbles over this.
There is a possible workaround with the Directus Seamless SSO Feature.

Seamless SSO
https://docs.directus.io/self-hosted/sso.html#seamless-sso

Flow  

1. Authenticate to Keycloak with the link mentioned in the Doc
`<a href="https://myDirectusInstance.com/auth/login/keycloak?redirect=https://myAppThatWantsAccessToDirectusAPI.com/">Login</a>`
2. Fetch a New Access Token from Directus with the Cookie you got from Keycloak
`await fetch('https://directus.myserver.com/auth/refresh', {
	method: 'POST',
	credentials: 'include', // this is required in order to send the refresh token cookie
});`
3. Call the Directus API with the newly generated Token from the Fetch-Response
`let resonseObject = await refreshResponse.json();
let newToken: string = resonseObject?.data.access_token;
requestHeaders.append("authorization", `bearer ${newToken}`);`

And your Requests will work.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
